enterprise/wsproxy/keycache.go

```diff
@@ -0,0 +1,252 @@
+	}
+
+	key, ok = keys[sequence]
+	if !ok {
+		return codersdk.CryptoKey{}, cryptokeys.ErrKeyNotFound
+	}
+
+	return validKey(key, now)
+}
+
+// refresh fetches the keys from the control plane and updates the cache.
+func (k *CryptoKeyCache) refresh() {
+	if k.isClosed() {
+		return
+	}
+
+	now := k.Clock.Now("CryptoKeyCache", "refresh")
+	k.fetchLock.Lock()
+	defer k.fetchLock.Unlock()
+
+	if k.isClosed() {
+		return
+	}
+
+	k.keysMu.RLock()
+	lastFetch := k.lastFetch
+	k.keysMu.RUnlock()
+
+	// There's a window we must account for where the timer fires while a fetch
+	// is ongoing but prior to the timer getting reset. In this case we want to
+	// avoid double fetching.
+	if now.Sub(lastFetch) < time.Minute*10 {
+		return
+	}
+
+	_, _, err := k.fetch(k.refreshCtx)
+	if err != nil {
+		k.logger.Error(k.refreshCtx, "fetch crypto keys", slog.Error(err))
+		return
+	}
+}
+
+// cryptoKeys queries the control plane for the crypto keys.
+// Outside of initialization, this should only be called by fetch.
+func (k *CryptoKeyCache) cryptoKeys(ctx context.Context) (map[int32]codersdk.CryptoKey, codersdk.CryptoKey, error) {
+	keys, err := k.client.CryptoKeys(ctx)
+	if err != nil {
+		return nil, codersdk.CryptoKey{}, xerrors.Errorf("crypto keys: %w", err)
+	}
+	cache, latest := toKeyMap(keys.CryptoKeys, k.Clock.Now())
+	return cache, latest, nil
+}
+
+// fetch fetches the keys from the control plane and updates the cache. The fetchMu
+// must be held when calling this function to avoid multiple concurrent fetches.
+// The returned keys are safe to use without additional locking.
+func (k *CryptoKeyCache) fetch(ctx context.Context) (map[int32]codersdk.CryptoKey, codersdk.CryptoKey, error) {
+	keys, latest, err := k.cryptoKeys(ctx)
+	if err != nil {
+		return nil, codersdk.CryptoKey{}, xerrors.Errorf("fetch keys: %w", err)
+	}
+
+	if len(keys) == 0 {
+		return nil, codersdk.CryptoKey{}, cryptokeys.ErrKeyNotFound
+	}
+
+	now := k.Clock.Now()
+	if !latest.CanSign(now) {
+		return nil, codersdk.CryptoKey{}, cryptokeys.ErrKeyInvalid
+	}
+
+	k.keysMu.Lock()
+	defer k.keysMu.Unlock()
+
+	k.lastFetch = k.Clock.Now()
+	k.refresher.Reset(time.Minute * 10)
+	k.keys, k.latest = maps.Clone(keys), latest
+
+	return keys, latest, nil
+}
+
+func toKeyMap(keys []codersdk.CryptoKey, now time.Time) (map[int32]codersdk.CryptoKey, codersdk.CryptoKey) {
+	m := make(map[int32]codersdk.CryptoKey)
+	var latest codersdk.CryptoKey
+	for _, key := range keys {
+		m[key.Sequence] = key
+		if key.Sequence > latest.Sequence && key.CanSign(now) {
+			latest = key
+		}
+	}
+	return m, latest
+}
+
+func validKey(key codersdk.CryptoKey, now time.Time) (codersdk.CryptoKey, error) {
+	if !key.CanVerify(now) {
+		return codersdk.CryptoKey{}, cryptokeys.ErrKeyInvalid
+	}
+
+	return key, nil
+}
+
+func (k *CryptoKeyCache) isClosed() bool {
+	return k.closed.Load()
+}
+
+func (k *CryptoKeyCache) Close() {
+	if k.isClosed() {
+		return
+	}
+
+	k.refreshCancel()
+	k.closed.Store(true)
```

Comment:
Multiple calls to `Close()` could race such that they both execute.  Right now `Close()` is idempotent, but I think it's better to just make `closed` a regular boolean protected by the mutex, so we don't have to worry about it getting executed more than once.